envoy/upstream/load_balancer.h

```diff
@@ -28,6 +28,35 @@ namespace Upstream {
 
 using ClusterProto = envoy::config::cluster::v3::Cluster;
 
+/*
+ * A handle to allow cancelation of asynchronous host selection.
+ * If chooseHost returns a HostSelectionResponse with an AsyncHostSelection
+ * handle, and the endpoint does not wish to receive onAsyncHostSelction call,
+ * it must call cancel() on the provided handle.
+ */
+struct AsyncHostSelection {
+  virtual ~AsyncHostSelection() = default;
+  virtual void cancel() PURE;
+};
+
+/*
+ * The response to a LoadBalancer::chooseHost call.
+ *
+ * chooseHost either returns a host directly or, in the case of asynchronous
+ * load balancing, returns an AsyncHostSelection handle.
+ *
+ * If it returns a AsyncHostSelection handle, the load balancer guarantees an
+ * eventual call to LoadBalancerContext::onAsyncHostSelction unless
+ * AsyncHostSelection::cancel is called.
+ */
+struct HostSelectionResponse {
+  HostSelectionResponse(HostConstSharedPtr host,
+                        std::shared_ptr<AsyncHostSelection> cancelable = nullptr)
+      : host(host), cancelable(std::move(cancelable)) {}
+  HostConstSharedPtr host;
+  std::shared_ptr<AsyncHostSelection> cancelable;
```

Comment:
OOC, why a shared_ptr here? Do we copy `HostSelectionResponse`, I guess? Perhaps we should use `unique_ptr<HostSelectionResponse>` instead and have `cacnelable` be an `optional` or something?

Reply:
so for one we have two different points possible in two different threads which can do things do this.  I could arguably do a unique pointer but then we have to release() whenever we cancel() in the router lest the other point have memory swiped, and that seemed more fragile

Also it turns out it's really hard to get the test mocks to handle moving things around.  I had byMove in like 50 tests but then it turns out you can't do default Returns because you can't ByMove something twice, so have to then do full function implementations of everything.  So I gave up and shared_ptr rather than balloon tests edits further.  I'm open to revisiting though.

Reply:
Following up from the offline discussion, I think we can make this `unique_ptr<Handle>` in this PR and then in the followup we need to sort out how to make the lifetime of `Handle` differ from the ownership of the underlying asyn resolution (maybe by hiding a shared_ptr or maybe doing something more specific)